File: backend/data/crawlAll.js
const fs = require('fs');
const crawlChosunEdu = require('../crawl/chosunedu');
const crawlDhNews = require('../crawl/dhnews');
const crawlIncheonNews = require('../crawl/incheon');
const crawlKcce = require('../crawl/kcce');
const crawlKyosu = require('../crawl/kyosu');
const crawlMoe = require('../crawl/moe');
const crawlUnipress = require('../crawl/unipress');
const crawlUnn = require('../crawl/unn');
const crawlUsline = require('../crawl/usline');
const crawlVeritas = require('../crawl/veritas-a');
const crawlYna = require('../crawl/yna');

async function crawlAllNews() {
    try {
      // 각 크롤러를 병렬로 실행
      const results = await Promise.all([
        crawlChosunEdu(),
        crawlDhNews(),
        crawlIncheonNews(),
        crawlKcce(),
        crawlKyosu(),
        crawlMoe(),
        crawlUnipress(),
        crawlUnn(),
        crawlUsline(),
        crawlVeritas(),
        crawlYna()
      ]);
  
      // 결과를 하나로 합치기
      const mergedResults = results.flat();
  
      // 결과를 JSON 파일로 저장
      fs.writeFileSync('news.json', JSON.stringify(mergedResults, null, 2));
  
      console.log('크롤링이 완료되었습니다. news.json 파일에 저장되었습니다.');
  
    } catch (error) {
      console.error('크롤링 중 오류 발생:', error);
    }
  }
  
  crawlAllNews();
  

crawlAllNews();

File: backend/crawlAll.py
import json
import threading
from crawl.chosun import get_data as chosun
from crawl.dhnews import get_data as dhnews
from crawl.incheon import get_data as incheon
from crawl.kcce import get_data as kcce
from crawl.kyosu import get_data as kyosu
from crawl.moe import get_data as moe
from crawl.unipress import get_data as unipress
from crawl.unn import get_data as unn
from crawl.usline import get_data as usline
from crawl.veritas import get_data as veritas
from crawl.yna import get_data as yna
from crawl.incheonilbo import get_data as incheonilbo
from crawl.kyeongin import get_data as kyeongin


lock = threading.Lock()

def fetch_news(source, result_list):
    data = source()
    if isinstance(data, list) and data != ['error']:
        with lock:
            result_list.extend(data)

def crawl_all():
    data = []
    threads = []

    sources = [chosun, dhnews, incheon, kcce, kyosu, moe, unipress, unn, usline, veritas, yna, incheonilbo, kyeongin]

    for source in sources:
        thread = threading.Thread(target=fetch_news, args=(source, data))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    # JSON 파일로 저장
    # with open("data/news.json", "w", encoding="utf-8") as f:
    #     json.dump(data, f, ensure_ascii=False, indent=4)

    print(f"크롤링 완료: 총 {len(data)}개 기사 수집")
    return data

# crawl_all()
